walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,67 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> busyIntervals = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> freeIntervals = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> importantRanges = new ArrayList<TimeRange>();
+    
+    for(String attendee : request.getAttendees()){
+      for (Event event : events){
+        if(event.getAttendees().contains(attendee)){
+          importantRanges.add(event.getWhen());
+        }
+      }
+    }
+
+    importantRanges.sort(TimeRange.ORDER_BY_START);
+   
+    TimeRange popedRange;
+    for(TimeRange timeRange : importantRanges){
+      if(busyIntervals.size() > 0 && timeRange.overlaps(busyIntervals.get(busyIntervals.size()-1))){
+        popedRange = busyIntervals.get(busyIntervals.size()-1);
+        busyIntervals.remove(busyIntervals.size() - 1);
+        busyIntervals.add(mergeTimeRanges(popedRange, timeRange));
+      }else{
+        busyIntervals.add(timeRange);
+      }
+    }
+
+    int duration;
+    if(busyIntervals.size() > 0 ) {
```

Comment:
If we add two dummy intervals one as (0,0) the other as (TimeRange.END_OF_DAY, TimeRange.END_OF_DAY), we'll be able to eliminate the start/end corner cases.

Reply:
Wow, this is really smart, thanks!